reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt

```diff
@@ -0,0 +1,626 @@
+        assertThat(context?.networkInfo?.carrierName).isNull()
+        assertThat(context?.deviceInfo?.deviceName).isNotEmpty()
+        assertThat(context?.deviceInfo?.deviceBrand).isNotEmpty()
+        assertThat(context?.deviceInfo?.deviceType).isEqualTo(DeviceType.MOBILE)
+        assertThat(context?.deviceInfo?.deviceModel).isNotEmpty()
+        assertThat(context?.time?.serverTimeNs).isGreaterThan(0)
+        assertThat(context?.time?.deviceTimeNs).isCloseTo(
+            TimeUnit.MILLISECONDS.toNanos(currentDeviceTimeInMs),
+            Offset.offset(DEVICE_TIME_OFFSET_NS)
+        )
+        assertThat(context?.userInfo?.id).isEqualTo(fakeUserId)
+        assertThat(context?.userInfo?.name).isEqualTo(fakeUserName)
+        assertThat(context?.userInfo?.email).isEqualTo(fakeUserEmail)
+        assertThat(context?.userInfo?.additionalProperties)
+            .containsExactlyInAnyOrderEntriesOf(fakeUserAdditionalProperties)
+        assertThat(context?.featuresContext).isEmpty()
+    }
+
+    @Test
+    fun mustReturnTheUpdatedFeatureContext_getDatadogContext_featureContextWasSet() {
+        // Given
+        val fakeKeyValues = forge.aMap { forge.anAlphabeticalString() to forge.anAlphabeticalString() }
+        featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+            fakeKeyValues.forEach { (key, value) ->
+                it[key] = value
+            }
+        }
+
+        // When
+        val context = testedInternalSdkCore?.getDatadogContext()
+
+        // Then
+        assertThat(context?.featuresContext?.entries).hasSize(1)
+        assertThat(context?.featuresContext?.get(fakeFeatureName)).containsExactlyInAnyOrderEntriesOf(fakeKeyValues)
+    }
+
+    @Test
+    fun mustUseAtomicOperations_when_updateFeatureContext_addNewValues() {
+        // Given
+        val fakeKeyValues1 = forge.aMap(
+            size = forge.anInt(
+                min = 1,
+                max = 10
+            )
+        ) { forge.anAlphabeticalString() to forge.anAlphabeticalString() }
+        val fakeKeyValues2 = forge.aMap(
+            size = forge.anInt(
+                min = 1,
+                max = 10
+            )
+        ) { forge.anAlphabeticalString() to forge.anAlphabeticalString() }
+        val expectedKeyValues = fakeKeyValues1 + fakeKeyValues2
+        val countDownLatch = CountDownLatch(2)
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues1.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues2.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        testedInternalSdkCore?.setContextUpdateReceiver(fakeFeatureName) { _, _ -> countDownLatch.countDown() }
```

Comment:
this should go before threads are started. There is a chance that some thread may finish its execution before this receiver is set.